### PR TITLE
Feature cn 107 prevent data loss

### PR DESCRIPTION
**Pull request by mandolyte**
# Describe what your pull request addresses

*This is integration work to use an unsaved data callback now available in the uw-editor component.*

- [ ] When a card is closed and there are unsaved changes to the book, then a dialog will appear
- [ ] This dialog will warn them and offer two choices, Continue (which closes the card) and Cancel (which returns the user to the card so they can click the save button).

## Test Instructions

- [ ] Open a book for which you have update permissions
- [ ] Make a change
- [ ] Click the "X" on the card to close 
- [ ] Observe that a dialog appears informing the user that they have unsaved changes
- [ ] If they click Cancel, the dialog simply disappears (then they can click save if they want to save it)
- [ ] If they click Continue, the dialog and the card will disappear
- [ ] Click to close the tab, a warning message is shown allowing the user to return
- [ ] Click to close the browser, a warning message is shown -- however, the app appears to be reset. Data may be lost?

NOTE: that if the book is re-opened, the data is there with the changes intact. However, the save button is not enabled. This is a bug? Workaround is to click the undo (which is enabled) and then redo. Then click save.

### CODE REVIEW

**Review by superdav42 — CHANGES_REQUESTED** [review]
Small change requested but it does not address all the requirements of the issue #107.
> Now close that tab or refresh the page

We need to show the same or similar warning when the users tries to close the tab with unsaved changes. The react hook `useBeforeUnload` should help tccreate does this with it https://github.com/unfoldingWord/tc-create-app/blob/c625b93417fe351ef08252a3c846386e09712553/src/Layout.js#L46

File: src/components/ScriptureWorkspace.js [hunk 1]
```diff
@@ -18,6 +18,9 @@ import NetworkErrorPopup from '@components/NetworkErrorPopUp'
 import ScriptureWorkspaceCard from './ScriptureWorkspaceCard'
 import useStoreContext from '@hooks/useStoreContext'
 import EmptyMessage from './EmptyMessage'
+import UnsavedDataPopup from './UnsavedDataPopup'
+import { data } from 'autoprefixer'
+import { SetMeal } from '@mui/icons-material'
```
[the remark this answers is not in the document]
Comment (mandolyte) [reply src/components/ScriptureWorkspace.js]: fixed

**Comment by mandolyte** [comment]
The updates address the tab closing, but closing the entire browser seems to work differently. I updated the test instructions. Let me know what you think.

I got the Nextjs example and pretty much pasted it into `/pages/_app.js`. I commented out the parts I did not think we needed, but left them there since the other conditions handled might prove useful.

File: src/components/Layout.js [hunk 2]
```diff
@@ -17,6 +20,64 @@ export default function Layout({ children, showChildren, title = APP_NAME }) {
...
+  //   }
+
+  //   if (contentIsDirty) {
+  //     event.preventDefault();
+  //     let _warning = 'There is unsaved data! Please confirm you wish to close the application';
+  //     event.returnValue = _warning
+  //     return _warning;
+  //   }
+  // });
+
```
Comment (superdav42) [inline src/components/Layout.js:80]: Removed this unused code. Looks like it's in _app.js
```suggestion
```

File: src/components/Layout.js [hunk 3]
```diff
@@ -1,9 +1,12 @@
 import { useContext, useEffect, useMemo, useRef, useState } from 'react'
 import PropTypes from 'prop-types'
+// import { useBeforeunload } from 'react-beforeunload';
```
Comment (superdav42) [inline src/components/Layout.js:3]:
```suggestion
```

File: src/components/Layout.js [hunk 4]
```diff
@@ -1,9 +1,12 @@
...
 import { AuthContext } from '@context/AuthContext'
 import Header from '@components/Header'
 import Footer from '@components/Footer'
 import { StoreContext } from '@context/StoreContext'
+import { AppContext } from '@context/AppContext'
+
```
Comment (superdav42) [inline src/components/Layout.js:9]:
```suggestion
```

**Review by superdav42 — CHANGES_REQUESTED** [review]
> Click to close the browser, a warning message is shown -- however, the app appears to be reset. Data may be lost?

It didn't work this way for me. Closing the app did the same thing as closing the tab. Nothing was loss

Remove the changes in Layout.js and I think this is golden.

**Pull request merged**